Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,315 @@
+
+For the second bucket in the spec shown in above example, the minimum replicas is `2`, the minimum cpu is `500m` (`1 replica * 1 core / 2 replicas`) and the minimum memory is `1G` (`1 replica * 2G memory / 2 replicas`).
+
+### kube-apiserver scaling example:
+For kube-apiserver, a typical spec could look like:
+
+```yaml
+spec:
+  scaleIntervals:
+  - maxCpu: "500m",
+    maxMemory: "2G",
+    maxReplicas: "1",
+  - maxCpu: "1",
+    maxMemory: "4G",
+    maxReplicas: "2",
+  - maxCpu: "2",
+    maxMemory: "8G",
+    maxReplicas: "3",
+  - maxCpu: "4",
+    maxMemory: "16G",
+    maxReplicas: "4",
+  - maxCpu: "8",
+    maxMemory: "32G",
+    maxReplicas: "5",
+  - maxCpu: "16",
+    maxMemory: "64G",
+    maxReplicas: "6",
+```
+* Here HVPA scales kube-apiserver from 1 to 2 replicas only when aggregate VPA recommendations cross 500m for CPU or 2GB for memory
+* Similarly kube-apiserver is scaled out from 2 replicas if aggregate VPA recommendations cross maximum aggregate resources for second bucket, i.e., 2 CPU or 8GB memory
+* Scale down from 3 replicas happen only when aggregate VPA recommendations go below minimum aggregate resources for third bucket, i.e., 2 CPU or 8GB memory. These values are obtained using maximum aggreagte resource values from second bucket
+
+**Scaling Example 1**: Scale Up & out:
+
+Current Requests: 150m / 1.8Gi with 1 replica
+VPA Recommendation: 150m / 2.2Gi (This puts us in 2nd bucket because of 2.2Gi aggregate memory recommendation)
+New Requests: 150m / 2.2Gi with 2 replicas, because 2.2Gi lands us in the “second” bucket when out-scaling that has 2 replicas; we carry over the vertical scaling as is, which is most conservative.
+Notice: New CPU would keep us in bucket 1 while memory brings us to bucket 2, so we go to bucket 2.
+Reasoning: We don’t know what the cache percentage is and if it is very high, the new replica will need almost the same memory to hold the cache. If not, VPA will eventually scale down.
+
+**Scaling Example 2**: Scale In & Down:
+
+Current Requests: 150m / 2.2Gi with 3 replicas
+VPA Recommendation: 150m / 1.8Gi (This puts us in 2nd bucket because of 5.4G aggregate memory recommendation)
+New Requests: 225m / 2.7Gi with 2 replicas, because 1.8Gi lands us in the “second” bucket when in-scaling that has 2 replicas; we adapt the vertical scaling based on the replica ratio, which is most conservative (3 * 1.8Gi = 5.4 Gi overall memory requests from the PoV of VPA; 5.4 Gi / 2 = 2.7 Gi memory requests per replica that preserves this value; CPU the same, i.e. 3 / 2 * 150m = 225m)
+Notice: Old and new CPU would even land us in bucket 1, but memory brings us only down to bucket 2, so we go to bucket 2 with 2 replicas.
+Reasoning: We don’t know what the cache percentage is and if it is very low, the remaining replicas will need even more memory to cope with the requests. If not, VPA will eventually scale down.
```

Comment:
> Reasoning: We don’t know what the cache percentage is and if it is very low, the remaining replicas will need even more memory to cope with the requests. If not, VPA will eventually scale down.

I am guessing the spirit of the statement here. Shouldn't it be _if it is very high_?

Reply:
Nope. In other words - since cache is supposed to be replicated among replicas, we don't need extra memory (per replica) for cache, but we might need that extra memory if the scaled down replica was handling lot of request.

Reply:
That doesn't make sense at all. If the high memory was because of mostly request processing, that usage is already seen by VPA and it is suggesting a scale down. Scaling (up or down) due to usage is handled equally well by scaling vertically or horizontally. Vertically or horizontally makes a difference only for if there is a significant memory foot-print of the cache.

Reply:
This statement is as per our discussion over email.
If there is another alternative to how we should scale in this case, please let me know what can be done in.
Thanks.

Reply:
> I am guessing the spirit of the statement here.

I am not questioning the logic of using VPA recommendation here. I was merely pointing out that even assuming the _spirit_ of the statement of the proposal (that it makes sense to use VPA recommendation while switching buckets during scale down) the stated rationale for using VPA recommendation (because of low memory foot-print of cache) doesn't make sense.

The pros and cons as well as alternatives for using VPA recommendation while switching buckets (during scale up or down) that were discussed in offline discussion, are not clear in the proposal and I can only comment after it is clearer in the proposal.

Reply:
>We don’t know what the cache percentage is and if it is very low, the remaining replicas will need even more memory to cope with the requests

I am not sure what is not clear in this statement. Do you think adding following statement would help?

"If the memory usage of the scaled down replica is mostly due to requests handling, then we would need to assign more memory to existing replicas to cope up with the requests"

Please feel free to suggest any edit to make it more understandable/sensible.
Thanks.